oca/api.py

```diff
@@ -53,45 +53,31 @@ def method(self):
             instantiated with an iterable of the decorated method results.
 
     Raises:
-        ValueError: If no container type is provided, and a value is returned
-            from the decorated method.
-        TypeError: If an incorrect data type is provided as the `container`.
-            Valid arguments are `None`, instances of `type`, or a callable.
+        AssertionError: If no container type is provided, and a value is
+            returned from the decorated method. This will only be raised if
+            ``PYTHONOPTIMIZE`` is less than 2.
+        AssertionError: If an incorrect data type is provided as `container`.
+            Valid arguments are `None` or instances of `type`. This will only
+            be raised if ``PYTHONOPTIMIZE`` is less than 2.
 
     Returns:
-        list: Of the return values from the decorated method, in the same
-            order as the original ``self`` RecordSet.
+        mixed: A new `container`, instantiated with the results of the
+            decorated method - iterated so that `self` is a singleton.
     """
 
-    method = None
-
-    if container is not None and not isinstance(container, type):
-        if callable(container):
-            method = container
-            container = None
-        else:
-            raise TypeError(
-                'An incorrect argument was provided to the `foreach` api.',
-            )
+    assert container is None or isinstance(container, type)
 
     def _foreach(method):
 
         def loop(method, self, *args, **kwargs):
-            results = [method(rec, *args, **kwargs) for rec in self]
+            results = (method(rec, *args, **kwargs) for rec in self)
             if container is None:
-                if not all(r is None for r in results):
-                    raise ValueError(
-                        'Methods decorated with `foreach` that return a '
-                        'value must declare the aggregation container type.',
-                    )
+                assert all(r is None for r in results)
```

Comment:
This command is what makes the generator be executed, which means it would never be in production.

Change it for:

```
if container is None:
    for result in results:  # Ensures the generator gets exhausted
        assert result is None  # Skips validation in production
```